db/transaction_log_impl.h

```diff
@@ -54,14 +56,80 @@ class WalFileImpl : public WalFile {
   uint64_t sizeFileBytes_;
 };
 
+class TransactionLogSeqCache {
+ public:
+  TransactionLogSeqCache(size_t size, SystemClock* clock)
+      : size_(size), clock_(clock) {}
+  struct SeqWithFileBlockIdx {
+    uint64_t log_number;
+    uint64_t seq_number;
+    uint64_t block_index;
+    uint64_t timestamp;
+    SeqWithFileBlockIdx(uint64_t log_num, uint64_t seq_num, uint64_t block_idx,
+                        uint64_t ts)
+        : log_number(log_num),
+          seq_number(seq_num),
+          block_index(block_idx),
```

Comment:
IMHO, log::kBlockSize should be internals of log reader and log writer. Since you are already converting offset to block index back and forth. How about just caching offset as opposed to block index. We can use more descriptive naming like write_batch_seqno, write_batch_offset show their logic relationship.